tempesta_fw/sock_srv.c

```diff
@@ -229,10 +237,6 @@ tfw_sock_srv_connect_try_later(TfwSrvConn *srv_conn)
 {
 	unsigned long timeout;
 
-	/* Don't rearm the reconnection timer if we're about to shutdown. */
-	if (unlikely(!ss_active()))
-		return;
```

Comment:
Why should we reconnect if we're about to shutdown? This may lead new reconnection/shutdown races.

Reply:
This check doesn't save us from reconnections on shutdown: failovering thread can pass this check before we set `false` for `__ss_active` during shutdown process, and we can get into `tfw_sock_srv_disconnect()` during live reconnection attempt. To prevent all these races - the waiting cycle had been introduced in `tfw_sock_srv_disconnect()`; so now if we missed connection during shutdown and it has time to begin `tfw_sock_srv_connect_try_later()`, it will still be stopped before reconnection attempt - due to successful timer deactivation in cycle [here](https://github.com/tempesta-tech/tempesta/blob/0fc370f5d3cc47df513dab7dbe214fd46eb12c8c/tempesta_fw/sock_srv.c#L479); in the rare worst case (which we cannot avoid in any way, in current architecture), when the timer handler is already started, and `del_timer_sync()` failed and return 0 (after handler finished), then three variants are possible:
1. Connection will be stopped [here](https://github.com/tempesta-tech/tempesta/blob/0fc370f5d3cc47df513dab7dbe214fd46eb12c8c/tempesta_fw/sock_srv.c#L151) (see [comment](https://github.com/tempesta-tech/tempesta/pull/1179/files#r255839826) above);
2. Connection destructor will be called [here](https://github.com/tempesta-tech/tempesta/blob/0fc370f5d3cc47df513dab7dbe214fd46eb12c8c/tempesta_fw/sock_srv.c#L201) and connection will be definitely stopped in it (since `TFW_CONN_B_DEL` is already set);
3. Connection is successfully restored, its reference count will became not equal `TFW_CONN_DEATHCNT` - then connection will be stopped in usual way, via `tfw_connection_close()` [here](https://github.com/tempesta-tech/tempesta/blob/0fc370f5d3cc47df513dab7dbe214fd46eb12c8c/tempesta_fw/sock_srv.c#L489).

In all these cases the race conditions are avoided, since we are waiting in `tfw_sock_srv_disconnect()` until connection will be stopped consistently before cleaning any of its resources.